fix: fiori 3 high contrast mode contrast ratios

Mike-Diaz opened the pull request:
## Related Issue
Closes SAP/fundamental-styles#1030
Closes SAP/fundamental-styles#1029
Closes SAP/fundamental-styles#1018
Closes SAP/fundamental-styles#1005
Closes SAP/fundamental-styles#998
Closes SAP/fundamental-styles#1038
Closes SAP/fundamental-styles#1033

## Description
This PR is focused on fixing Fiori High Contrast White and High Contrast Black issues with several components. Components currently broken, or violating WCAGAA contrast ratio guidelines have been addressed and now pass ratio recommendations.

#### New ratios:
Tested with PCG Colour Contrast Analyser

- `Avatar` Tile Icon Background Fiori HCW: 21:1
- `Avatar` Tile Icon Background Fiori HCB: 21:1
- `Avatar` Icon Fiori HCW: 7.3:1
- `Tabs` Default Fiori HCB: 11.7:1
- `Tabs` Icon Only Fiori HCB: 11.7:1
- `Tabs` Default Fiori HCW: 7.1:1
- `Tabs` Icon Only HCW: 7.1:1
- `Radio` Responsiveness Fiori HCB: 21:1
- `Radio` Responsiveness Fiori HCW: 7.3:1
- `Checkbox` Desktop Fiori HCB: 21:1
- `Checkbox` Desktop Fiori HCW: 7.3:1

#### Note:
`Checkbox` and `Radio` screenshots in IE11 are broken. Using CSS Addons to switch styles in Canvas does not work (in IE11). Using the Theme selector in Docs works partially. Theme selector only shows some of the styling changes, but it will not change the background color of the frame they are housed in, therefore; any theme using white colors for text or other elements will not be visible since the background is also white.

When testing these changes, please use Canvas instead of Docs to prevent the issues mentioned above with other components.

## Screenshots

### Before:
![Avatar_Before](https://user-images.githubusercontent.com/21978807/96166827-60366d00-0ed3-11eb-8985-b29106494dfb.png)
![Avatar_Icon_Before](https://user-images.githubusercontent.com/21978807/96166853-67f61180-0ed3-11eb-951a-dbff730e2978.png)
![Avatar_TileIcon_Before](https://user-images.githubusercontent.com/21978807/96166858-6a586b80-0ed3-11eb-8cfd-3f58e6ee6915.png)
![Radio_Before](https://user-images.githubusercontent.com/21978807/96166871-6d535c00-0ed3-11eb-8a28-dbd77e75208f.png)
![Radio_Before_HCW](https://user-images.githubusercontent.com/21978807/96312223-0495dd80-0fc0-11eb-99e2-edc10b8ae1a5.png)
![Checkbox_Before_HCW](https://user-images.githubusercontent.com/21978807/96312185-f34cd100-0fbf-11eb-8744-a6df8ca68856.png)
![Tabs_Before](https://user-images.githubusercontent.com/21978807/96166888-72b0a680-0ed3-11eb-827a-5bd37dffb3a0.png)

### After:
#### Chrome
<img width="1342" alt="Avatar_Icon_After" src="https://user-images.githubusercontent.com/21978807/96170084-fc627300-0ed7-11eb-80bc-c130c6703301.png">
<img width="1344" alt="Avatar_Placeholder_HCB" src="https://user-images.githubusercontent.com/21978807/96299153-6e09f200-0fa8-11eb-8831-47d0f7b0bee5.png">
<img width="1344" alt="Avatar_Placeholder_HCW" src="https://user-images.githubusercontent.com/21978807/96299161-72360f80-0fa8-11eb-98a3-12f95d2cf5b6.png">
<img width="1342" alt="Avatar_TileIcon_HCB" src="https://user-images.githubusercontent.com/21978807/96299468-fa1c1980-0fa8-11eb-872d-a8b7bfd0d7ae.png">
<img width="1342" alt="Avatar_TileIcon_HCW" src="https://user-images.githubusercontent.com/21978807/96299487-fe483700-0fa8-11eb-881b-2bedebb94d96.png">
<img width="1342" alt="Checkbox_Desktop_After" src="https://user-images.githubusercontent.com/21978807/96170119-0a17f880-0ed8-11eb-84a3-f5f2b986738c.png">
<img width="1342" alt="Checkbox_Desktop_HCW" src="https://user-images.githubusercontent.com/21978807/96312286-21caac00-0fc0-11eb-8597-e471c4a31eb8.png">
<img width="1342" alt="Radio_Responsive_After" src="https://user-images.githubusercontent.com/21978807/96170131-0edcac80-0ed8-11eb-97b2-0a9b2df36bf0.png">
<img width="1342" alt="Radio_Responsive_HCW" src="https://user-images.githubusercontent.com/21978807/96312307-2becaa80-0fc0-11eb-807c-3789b28fb30e.png">
<img width="1342" alt="Tabs_Counter_After" src="https://user-images.githubusercontent.com/21978807/96170143-11d79d00-0ed8-11eb-90a3-7c04e2b8047f.png">
<img width="1342" alt="Tabs_Default_AFter" src="https://user-images.githubusercontent.com/21978807/96170147-13a16080-0ed8-11eb-8021-877efe0e69b4.png">
<img width="1342" alt="Tabs_Icon_After" src="https://user-images.githubusercontent.com/21978807/96170152-156b2400-0ed8-11eb-9ee1-94cb62e1bbbb.png">
<img width="1342" alt="Tabs_Default_HCW" src="https://user-images.githubusercontent.com/21978807/96308607-8eda4380-0fb8-11eb-939b-d62f94baf7a2.png">
<img width="1342" alt="Tabs_Icon_HCW" src="https://user-images.githubusercontent.com/21978807/96308622-94378e00-0fb8-11eb-995d-0985b203e7d3.png">

#### IE11
![IE11_Avatar_TileIcon_HCB](https://user-images.githubusercontent.com/21978807/96301920-c4792f80-0fac-11eb-8bbe-3800f3de9626.JPG)
![IE11_Avatar_TileIcon_HCW](https://user-images.githubusercontent.com/21978807/96301933-c7742000-0fac-11eb-9165-4c255d19eeb7.JPG)
![IE11_Avatar_Placeholder_HCB](https://user-images.githubusercontent.com/21978807/96301948-cc38d400-0fac-11eb-8110-8d05b3da60de.JPG)
![IE11_Avatar_Placeholder_HCW](https://user-images.githubusercontent.com/21978807/96301963-d064f180-0fac-11eb-99da-2057430ec179.JPG)
![IE11_Icon_After](https://user-images.githubusercontent.com/21978807/96170289-42b7d200-0ed8-11eb-8a96-7bf3e4268ede.JPG)
![IE11_Checkbox_After](https://user-images.githubusercontent.com/21978807/96170308-477c8600-0ed8-11eb-9012-ed7c0b08bcd2.JPG)
![IE11_Checkbox_HCW](https://user-images.githubusercontent.com/21978807/96315234-12992d80-0fc3-11eb-896a-71bcaf0d05f4.JPG)
![IE11_Radio_After](https://user-images.githubusercontent.com/21978807/96170319-49dee000-0ed8-11eb-9215-c2247d6a9c60.JPG)
![IE11_Radio_HCW](https://user-images.githubusercontent.com/21978807/96314848-f39a9b80-0fc2-11eb-8c2f-0232fc67bea9.JPG)
![IE11_Tabs_After](https://user-images.githubusercontent.com/21978807/96170325-4c413a00-0ed8-11eb-95a2-4cc0879f0578.JPG)
![IE11_Tabs_Icon_After](https://user-images.githubusercontent.com/21978807/96170333-4e0afd80-0ed8-11eb-8fd8-b68dc615af79.JPG)
![IE11_Tabs_Default_HCW](https://user-images.githubusercontent.com/21978807/96309980-3bb5c000-0fbb-11eb-9ad2-5a3946df4199.JPG)
![IE11_Tabs_Icon_HCW](https://user-images.githubusercontent.com/21978807/96309999-46705500-0fbb-11eb-88d7-c1ea7331d81f.JPG)

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [x] hover state of the element follow design spec
- [x] focus state of the element follow design spec
- [x] active state of the element follow design spec
- [x] selected state of the element follow design spec
- [x] selected hover state of the element follow design spec
- [x] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [x] BEM naming convention is used
- [x] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [x] `fd-reset()` mixin is applied to all elements
- [x] Variables are used, if some value is used more than twice.
- [x] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] tested Storybook examples with "CSS Resources" `normalize` option 
- [x] tested Storybook examples with "CSS Resources" `unnormalize` option 
- [x] Verified all styles in IE11
- [x] Updated tests
4. Documentation
- [x] Storybook documentation has been created/updated
- [x] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

prsdthkr commented:
Hi @mike, could you merge the master branch into this so that the new visual snapshots be generated in Chromatic?

prsdthkr reviewed (changes requested):
Avatar > Tile Icon Background in High Contrast Light:
![image](https://user-images.githubusercontent.com/43764832/96192065-2036b080-0efa-11eb-9a0f-4783da715a2c.png)

prsdthkr commented:
Can we also fix https://github.com/SAP/fundamental-styles/issues/1038 as part of this change?

jbadan commented:
Did any of these changes deviate from the published spec on the wiki? If so, we should email the designer responsible and let them know. If not, carry on!

prsdthkr commented:
Contrast for Avatar with a white icon/text on cyan(ish) bg for HCD theme isn't enough:
![image](https://user-images.githubusercontent.com/43764832/96327361-aaac0c80-0fed-11eb-937c-5dac7f87852f.png)

****

prsdthkr commented:
Avatar with borders has no borders in the Dark theme
![image](https://user-images.githubusercontent.com/43764832/96327649-4474b900-0ff0-11eb-9d6d-9b2714ec24e7.png)

Mike-Diaz commented:
@prsdthkr 

> Contrast for Avatar with a white icon/text on cyan(ish) bg for HCD theme isn't enough:

Changed the inner icon color to black.
<img width="1342" alt="Screen Shot 2020-10-18 at 2 44 06 PM" src="https://user-images.githubusercontent.com/21978807/96386562-bcf28b80-1150-11eb-8383-62a076fd5c45.png">

> Avatar with borders has no borders in the Dark theme

Oops, I messed this one up when I was adding borders to HCB elements. Added a new variable for the optional border element to use.
<img width="1342" alt="Screen Shot 2020-10-18 at 2 43 32 PM" src="https://user-images.githubusercontent.com/21978807/96386544-9e8c9000-1150-11eb-9bb4-46f656363723.png">

jbadan commented on src/avatar.scss:
```diff
@@ -40,7 +40,7 @@ $block: #{$fd-namespace}-avatar;
       @include fd-reset();
 
-      color: var(--sapContent_ContrastTextColor);
+      color: var(--fdAvatar_DefaultColor);
```
I think this should stay ` --sapContent_ContrastTextColor`. It's specified for all themes for the default color - this should be white in hcb and black in hcw. We've got the opposite happening.

jbadan commented on src/avatar.scss:
```diff
@@ -40,7 +40,7 @@ $block: #{$fd-namespace}-avatar;
 
-      color: var(--sapContent_ContrastTextColor);
+      color: var(--fdAvatar_DefaultColor);
       background-color: $fd-avatar-background-color;
```
`$fd-avatar-background-color` needs a new variable - it should be `--sapAccentColor6` in light + dark, `--sapBackgroundColor` in hcw/hcb.

jbadan commented on src/avatar.scss:
```diff
@@ -97,17 +97,21 @@ $block: #{$fd-namespace}-avatar;
   // optional border
   &--border {
-    border: 0.0625rem solid var(--sapGroup_ContentBorderColor);
+    border: var(--fdAvatar_Optional_Border);
```
This one can just be `0.0625rem solid var(--sapGroup_ContentBorderColor);` because it's the same in all themes. We can delete `--fdAvatar_Optional_Border`

jbadan commented on src/avatar.scss:
```diff
@@ -97,17 +97,21 @@ $block: #{$fd-namespace}-avatar;
   // placeholder background
   &--placeholder {
-    background-color: var(--sapContent_ImagePlaceholderBackground);
+    background-color: var(--fdAvatar_Placeholder_BackgroundColor);
```
This one can just be `var(--sapContent_ImagePlaceholderBackground);` and delete `--fdAvatar_Placeholder_BackgroundColor`

jbadan commented on src/avatar.scss:
```diff
@@ -97,17 +97,21 @@ $block: #{$fd-namespace}-avatar;
   &--placeholder {
-    background-color: var(--sapContent_ImagePlaceholderBackground);
+    background-color: var(--fdAvatar_Placeholder_BackgroundColor);
+    border: var(--fdAvatar_Border);
```
I think we can get rid of this border here and apply it to the root as it's the same for default.

jbadan commented on src/avatar.scss:
```diff
@@ -97,17 +97,21 @@ $block: #{$fd-namespace}-avatar;
   &--tile {
-    background-color: var(--sapTile_IconColor);
+    background-color: var(--fdAvatar_Tile_BackgroundColor);
+    border: var(--fdAvatar_Border);
```
I think we can get rid of this border here and apply it to the root as it's the same for default.

jbadan commented on src/avatar.scss:
```diff
@@ -97,17 +97,21 @@ $block: #{$fd-namespace}-avatar;
-    background-color: var(--sapContent_ImagePlaceholderBackground);
+    background-color: var(--fdAvatar_Placeholder_BackgroundColor);
+    border: var(--fdAvatar_Border);
+    color: var(--fdAvatar_Placeholder_DefaultColor);
```
We can also remove this color - I think sorting the default background out will actually solve most of the color contrast issues.

jbadan commented:
I pushed a couple changes for Avatar - feel free to revert if you don't like it :) 

![Screen Shot 2020-10-19 at 3 53 45 PM](https://user-images.githubusercontent.com/29607818/96520212-728e0f00-1223-11eb-84c9-207ab4c8e886.png)
![Screen Shot 2020-10-19 at 3 53 51 PM](https://user-images.githubusercontent.com/29607818/96520216-7326a580-1223-11eb-929b-de56dddc4d1c.png)
![Screen Shot 2020-10-19 at 3 53 57 PM](https://user-images.githubusercontent.com/29607818/96520220-73bf3c00-1223-11eb-8f27-0eaacd18e19c.png)
![Screen Shot 2020-10-19 at 3 54 01 PM](https://user-images.githubusercontent.com/29607818/96520221-7457d280-1223-11eb-94da-64e6269b22b3.png)
![Screen Shot 2020-10-19 at 2 58 22 PM](https://user-images.githubusercontent.com/29607818/96520222-7457d280-1223-11eb-9229-21e86ade51a1.png)

JKMarkowski reviewed (commented):
Hello @Mike-Diaz Is this PR only about colors? If so I will approve it. I see checkbox should have different border style on HCB on some states.
![image](https://user-images.githubusercontent.com/26483208/96571792-58f4d200-12cc-11eb-938f-b999fa33173d.png)

jbadan commented:
@JKMarkowski I can't speak for @Mike-Diaz, but I was ignoring that one for the moment. I think we need to do a massive audit and use 1 variable for border style. That same `solid` versus `dashed` appears in the majority of components. I'd hate to do it piecemeal and miss components or end up with 20 variables that are the same thing. 

Maybe in a followup pr we create some shared component variables `--fd-border` or whatever naming convention we want to use for these.

Mike-Diaz commented:
> Hello @Mike-Diaz Is this PR only about colors? If so I will approve it. I see checkbox should have different border style on HCB on some states.
> ![image](https://user-images.githubusercontent.com/26483208/96571792-58f4d200-12cc-11eb-938f-b999fa33173d.png)

@JKMarkowski Thanks for taking a look at the PR! Yes, this PR was mainly to deal with WCAG AA high contrast color issues. There is certainly more work that can be done, but I would like to keep the scope of the PR to only the issues mentioned in the description, if possible, so that it can eventually be merged in haha!

prsdthkr commented:
Tabs, Checkbox, and Radio LGTM.

For accented Avatars, 
Seems like an icon with a black color would offer better contrast in the HCD theme:
![image](https://user-images.githubusercontent.com/43764832/96643245-24891200-12dc-11eb-9114-c99ec40daf9c.png)

Seems like an icon with a white color would offer better contrast in the HCL theme:
![image](https://user-images.githubusercontent.com/43764832/96643321-3ff41d00-12dc-11eb-81ce-d7266f7bb4a6.png)

prsdthkr commented:
Also, could you merge master into this again for snapshot consistency?

Mike-Diaz commented:
> For accented Avatars,
> Seems like an icon with a black color would offer better contrast in the HCD theme:

> Seems like an icon with a white color would offer better contrast in the HCL theme:

Ah yes, I did notice this. I didn't pay much attention since this was not part of the issues addressed in this PR, but I went ahead and changed the inner icon color as requested :)

Also merged master in so that the snapshots are consistent.

prsdthkr reviewed (approved):
QA LGTM 🚢 
Thanks for the extra accent fix 👍
